Fix health event done

Farice4 opened the pull request.

Farice4 committed:
Fix health manager event listener thread_done error

Bug-ES #11210
http://192.168.15.2/issues/11210

Signed-off-by: Yuanbin.Chen <cybing4@gmail.com>

Farice4 committed:
Fix health detach policy event listener messaging not stop

Bug-ES #11211
http://192.168.15.2/issues/11211

Signed-off-by: Yuanbin.Chen <cybing4@gmail.com>

zhaochao commented on senlin/engine/health_manager.py:
```diff
@@ -266,7 +269,13 @@ def _stop_check(self, entry):
+                if cluster_id == lmsg['cluster_id']:
+                    lmsg['listener'].stop()
+                    lmsg['listener'].wait()
+                    Listeners.pop(Listeners.index(lmsg))
```
在上游的 gerrit 上我加了一个评论，大致的意思是没有必要单独维护个全局的记录，可以让 `ListenerProc` 把底层 olso.messaging 的 listener 返回，然后在 _stop_check 的时候取出这个 listener ，直接进行 stop 和 wait 。

Farice4 replied:
self.TG.add_thread(ListenerProc, 'nova', project, cluster_id) 这里的add_thread在oslo.service里面已经有一个返回值了,返回的是thread 值, oslo.messaging start的时候是一个返回值. 两个不同的返回值,在做stop的时候,是没有办法oslo.messaging stop的.

在这个过程中return 的oslo.messaging start,发生了变化, 只能单独存储这个返回值.

zhaochao replied:
ListenerProc 目前没有返回值，要按我说的方法改的话，直接返回 listener （这个是 oslo.messaging 的 notification listener 对象）。这个对象不会在 add_thread 时返回，只能通过 GreedThread.wait() 来获取。

至于 _stop_check 里的改动，我在上游 gerrit 里给了例子：
```
        if listener:
            messaging_listener = listener.wait()
            messaging_listener.stop()
            messaging_listener.wait()
            listener.stop()
```
 
不过我在上游 gerrit 那里又增加了一个 comment ，按照 `RUIJIE YUAN` 在 bug 描述里的建议直接在当前线程启动 oslo.messaging notification listener 似乎更好一点，因为这个本来就是单独启了一个线程，没必要再启动一个 greanthread 加到 threadgroup 里去。

fabian4 commented:
先不合并,看起来还有点问题.

blkart commented:
不合并的 pr 可以在标题前加上 WIP:

2017-11-23 9:23 GMT+08:00 chenyb4 <notifications@github.com>:

> 先不合并,看起来还有点问题.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/eayunstack/senlin/pull/83#issuecomment-346513838>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AG-IQbV6uOYVV-GrSNpUw-o1FyAlDtwAks5s5MkAgaJpZM4Qmz2m>
> .
>



-- 
-----------------------------------------------8<-----------------------------------------------------------------------------------------------

http://blkart.org

————Stay hungry. Stay foolish.

zhaochao reviewed (changes requested):
参考上游 gerrit 上的意见，在 HealthManager.stop() 这里的确也需要停止所有已注册的 listener 。

Farice4 commented:
done

Farice4 committed:
Fix health manager node recover twice

This patch fix listener nova delete operation,
receive "shutdown.end/delete.end" message node
recover operation repeat.

Bug-ES #11293
http://192.168.15.2/issues/11293

Signed-off-by: Yuanbin.Chen <cybing4@gmail.com>

chenpengzi commented on senlin/engine/health_manager.py:
```diff
@@ -449,3 +454,16 @@ def get_manager_engine(cluster_id):
+        if event_listener in Listeners:
+            Listeners.remove(event_listener)
+    else:
+        if len(Listeners):
```
直接if Listeners: 等同于if len(Listeners)

Farice4 replied:
Done
